### PR TITLE
feat: discovery peer filtering for relay shard

### PR DESCRIPTION
**Pull request by SionoiS**
# Description

Filter peer discovered by static shard information

# Changes

- added predicate based on relay shard info.
- apply predicate when peers are discovered
- test predicate

Step 3 of #1776 

TODO: ~~test predicate proc~~,

### CODE REVIEW

**Comment by SionoiS** [comment]
Hopefully we gucci this time...

File: waku/v2/waku_discv5.nim [hunk 1]
```diff
@@ -170,15 +170,37 @@ proc closeWait*(wd: WakuDiscoveryV5) {.async.} =
...
+      debug "no relay sharding information, peer filtering disabled"
+      return none(WakuDiscv5Predicate)
+    else: nodeShardOp.get()
+
+  debug "peer filtering enabled"
+
+  return some(proc(record: waku_enr.Record): bool =
+      nodeShard.indices.anyIt(record.containsShard(nodeShard.cluster, it)))
+
+proc findRandomPeers*(wd: WakuDiscoveryV5, pred: Option[WakuDiscv5Predicate] = none(WakuDiscv5Predicate)): Future[seq[waku_enr.Record]] {.async.} =
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:195]: Where a `type` (for an argument, initialised variable, etc.) is implicitly clear, it's usually not necessary to be overly explicit and IMO makes the code a bit more readable. Up to your own style and preference, though.

```suggestion
proc findRandomPeers*(wd: WakuDiscoveryV5, pred = none(WakuDiscv5Predicate)): Future[seq[waku_enr.Record]] {.async.} =
```

File: waku/v2/waku_discv5.nim [hunk 2]
```diff
@@ -170,15 +170,37 @@ proc closeWait*(wd: WakuDiscoveryV5) {.async.} =
...
+  let nodeShard =
+    if nodeShardOp.isNone():
+      debug "no relay sharding information, peer filtering disabled"
+      return none(WakuDiscv5Predicate)
+    else: nodeShardOp.get()
+
+  debug "peer filtering enabled"
+
+  return some(proc(record: waku_enr.Record): bool =
+      nodeShard.indices.anyIt(record.containsShard(nodeShard.cluster, it)))
```
Comment (jm-clius) [inline waku/v2/waku_discv5.nim:193]: Nitpick: WDYT about pulling the pred proc out here in a separate `let pred =` statement to make the `return` more readable? Something like:

```suggestion
  let pred = proc(record: waku_enr.Record): bool =
    nodeShard.indices.anyIt(record.containsShard(nodeShard.cluster, it))
  return some(pred)
```

File: tests/v2/test_waku_discv5.nim [hunk 3]
```diff
@@ -309,3 +309,77 @@ procSuite "Waku Discovery v5":
...
+
+      let shards = RelayShards.init(shardCluster, shardIndices)
+
+      var builder = EnrBuilder.init(enrPrivKey, seqNum = enrSeqNum)
+      require builder.withWakuRelaySharding(shards).isOk()
+
+      let recordRes = builder.build()
+      require recordRes.isOk()
+      recordRes.tryGet()
+
```
Comment (jm-clius) [inline tests/v2/test_waku_discv5.nim:367]: `#When` ? (Trying to follow the GIVEN-WHEN-THEN pattern and missing the "When" :D )

**Review by jm-clius — APPROVED** [review]
LGTM! Great to see. The comments below are only nitpicks or style suggestions, so feel free to implement as you see fit. Next steps are: configuring a (new) fleet with static sharding and (in parallel?) considering moving discv5 out of the `waku_node`.

File: waku/v2/waku_discv5.nim (hunk 1 above)
Comment (SionoiS) [reply waku/v2/waku_discv5.nim:195]: Ah good to know!

**Pull request merged**